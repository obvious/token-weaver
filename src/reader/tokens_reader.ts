import {readFile, writeFile, mkdir} from 'node:fs/promises';
import * as path from 'path';
import {existsSync} from 'fs';
import {Theme} from '../models/theme';

const fse = require('fs-extra');

const TEMP_TOKENS_FOLDER = 'temp_tokens';

const themeName = (theme: Theme) => `theme/${theme.name}`;
export const tempTokensDirPath = () =>
  path.join(__dirname, '..', TEMP_TOKENS_FOLDER);

export async function readTokens(tokensPath: string): Promise<{
  coreTokensPath: string;
  themeTokensPaths: string[];
}> {
  if (hasSingleTokensFile(tokensPath)) {
    await splitTokensFile(tokensPath);
  } else {
    await copyToTempDir(tokensPath);
  }

  const tokensDirPath = tempTokensDirPath();
  const themes: Theme[] = JSON.parse(
    await readFile(path.join(tokensDirPath, '$themes.json'), {
      encoding: 'utf-8',
      flag: 'r',
    })
  );

  return {
    coreTokensPath: path.join(tokensDirPath, 'core.json'),
    themeTokensPaths: themes.map(theme => {
      return path.join(tokensDirPath, `${themeName(theme)}.json`);
    }),
  };
}

async function copyToTempDir(tokensPath: string) {
  fse.copySync(tokensPath, tempTokensDirPath(), {overwrite: true});
}

async function splitTokensFile(tokensPath: string) {
  const tokensFile = await readFile(tokensPath, {encoding: 'utf-8', flag: 'r'});
  const tokens = JSON.parse(tokensFile);

  const core = tokens['core'];
  const themesList: Theme[] = tokens['$themes'];

  await writeJson('core', core);
  await writeJson('$themes', themesList);

  for (const theme of themesList) {
    const name = themeName(theme);
    const themeData = tokens[name];

    await writeJson(name, themeData);
  }
}

async function writeJson(fileName: string, content: any) {
  const filePath = path.join(tempTokensDirPath(), `${fileName}.json`);
  const dirPath = path.dirname(filePath);

  if (!existsSync(dirPath)) {
    await mkdir(dirPath, {recursive: true});
  }

  await writeFile(filePath, JSON.stringify(content));
}

function hasSingleTokensFile(tokensPath: string): boolean {
  return tokensPath.endsWith('.json');
}
