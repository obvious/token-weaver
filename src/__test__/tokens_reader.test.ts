import {readTokens, tempTokensDirPath} from '../reader/tokens_reader';
import {rm} from 'fs/promises';
import * as path from 'path';

describe('read tokens', () => {
  it('should read tokens from a single file', async () => {
    // given
    const tokensPath = path.join(
      __dirname,
      '../..',
      'sample_tokens',
      'tokens.json'
    );

    // when
    const result = await readTokens(tokensPath);

    // then
    expect(result.coreTokensPath).toBeTruthy();
  });

  it('should read tokens from multiple files', async () => {
    // given
    const tokensPath = path.join(__dirname, '../..', 'sample_tokens');

    // when
    const result = await readTokens(tokensPath);

    // then
    expect(result.themeTokensPaths).toBeTruthy();
    expect(result.themeTokensPaths.length > 0).toBeTruthy();
  });

  afterAll(async () => {
    await rm(tempTokensDirPath(), {recursive: true});
  });
});
