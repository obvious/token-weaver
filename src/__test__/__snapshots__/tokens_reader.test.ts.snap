// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`read tokens should read tokens from a single file 1`] = `
{
  "coreTokensPath": "/Users/sasikanth/Projects/weaver/src/temp_tokens/core.json",
  "themeTokensPaths": [
    "/Users/sasikanth/Projects/weaver/src/temp_tokens/theme/default.json",
    "/Users/sasikanth/Projects/weaver/src/temp_tokens/theme/dark.json",
  ],
}
`;

exports[`read tokens should read tokens from multiple files 1`] = `
{
  "coreTokensPath": "/Users/sasikanth/Projects/weaver/src/temp_tokens/core.json",
  "themeTokensPaths": [
    "/Users/sasikanth/Projects/weaver/src/temp_tokens/theme/default.json",
    "/Users/sasikanth/Projects/weaver/src/temp_tokens/theme/dark.json",
  ],
}
`;
